### ISSUE TITLE
Migrate web app to standalone executable

### ISSUE THREAD

**Issue by TelluriumCrystal** [TelluriumCrystal/ocap-revived#6]
It doesn't make sense for this sort of program to be a web app since it doesn't need to be hosted remotely. There exist ways to run javascript in an executable, so OCAP should be migrated to such a system.

**Comment by schbaeggle** [comment]
I would like to start a discussion based of that, if you allow.  Maybe I misunderstood that title, then pardon me ;)

In my opinion, a webapp - like a legacy version had - excels for platform independent usage as well as getting rid of bloatware filling your computer. The modern approach to systems like that - especially if some thoughts towards collaboration are given - is to make them web based. Zero install, can be viewed by a wide range of devices and OS. Having a short discussion of your mission afterwards by sending a URL and a timestamp (or even syncing timestamps across a "debrief group" like) suits quite a bit better than opening some application on a desktop.

Given my experience, we all most likely use a web-based system for our groups/clans like a forum, CMS, whiteboard, mission planner or even some mods use them. After all, just opening another tab for OCAP is quite good, I guess. Some users may not be as technically robust with all that "computer stuff", sending a URL to them prevents a lot of supportive work inside clans.

Plus, having two servers communicate all the data ONCE should be more performant compared to a scenario where every client either downloads all the date at once or even scrubbing through the timeline and demanding additional downloads all the time.

Having a web app - in my case - decides whether I will use OCAP or not.

Please do not understand that post as a complaint, just want to share my thoughts.

**Comment by Axodah** [comment]
Part of the intent of moving to a standalone application is because we are not intending to host an always-on server to support users of OCAP. The original application used that paradigm, and once abandoned by its creator(s), the application was no longer functional. Our goal with OCAP Revived is to complete some functionality that OCAP never got around to implementing, while allowing others to use it with or without our continued support. Converting this application into a platform independent standalone app allows it to be used by anyone at any point in the future, and it can still be used by groups (using outside tools like Discord screen sharing, so only one user would need to run the app). If there are other ways to achieve the ease of access that you would like, I would be happy to look into it.

We are open to functionality suggestions, and appreciate your feedback!

**Comment by TelluriumCrystal** [comment]
@Axodah hit the nail on the head. We do not have any plans to develop OCAP Revived to run as a web app. This is both because we do not want to introduce any dependency on a remote server, and because we want it to be very simple to use. A good example of the type of program we are ultimately aiming for is [TacView](https://www.tacview.net/) (but 2D instead of 3D).

That being said (and @Axodah can correct me if I'm wrong here) our implementation of the standalone app is essentially going to be an executable encapsulating a local server that OCAP will run on. So if you want to, it shouldn't be more complex than moving some files to install it on a web server.

We _might_ (and that's a big might) be inclined to release a version intended to run on a server in the far future. But before we can even consider that we need to finish implementing basic features and get everything working reliably.

We do appreciate discussion like this though and by all means feel free to give any feedback you wish.

**Comment by schbaeggle** [comment]
Sure, I'd say we all agree on: that always-on server is quite stupid and the main reason why OCAP needs a revival ;)
But everything can be packed into a single web-app (like a3 map data, database etc.) and offered for download to be self-hosted. So no central server is needed.

In my (quite small) brain, I can't imagine a more simple way than having that a3 server directly connecting to a webserver and pushing its data onto it. That webserver accesses his local database and renders all that fancy stuff you guys are planning. Everything has static addresses and ports, once set up, you can actually forget about it and enjoy your mission playbacks. Zero afford once set up (well, let's assume BI does not break any scripts while updating arma ;))

Having it a locally running app, all clients either need to know the ftp-connection credentials to download any kind of data or the a3 server needs to push its data to local clients (having the need of open ports, NAT settings, bla bla bla). To be honest, I can't see your point, @Axodah, of a packed standalone webapp running locally being platform independent. (In terms of being used by a default user). Furthermore, if it's actually an incapsulated webserver, why bother publishing it as a native web app and incapsulating it afterwards for local use?

Sure thing, both, a local app as well as a hosted webapp are doing the same work. But if I'm thinking about your target audience, I'd imagine it suites more to have a single URL to send to all players for debrief, maybe even having a feature to sync your playback across all your players. pressing pause and having the possibility to draw onto a map (like you often can see on sports-tv analysis) can be a huge feature to have. 
(As a lot of groups are either using TFAR or ACRE, we're bound to TS3 instead of discord, so using discords screen sharing is not as easy and common as you might think).

Maybe I don't get the tech-background and hurdles you are facing, but I do get the need of you guys reviving the only mod out there doing a simple task: record my mission and play them back (please don't name ALiVE or any other boatware here ;)). In the past, we as a mil-sim-based group used OCAP and it helped us enhance our tactics. I miss it and really can't wait to use it again ;)

**Comment by Axodah** [comment]
Ah, I see where you are coming from. If I correctly interpreted the original way that OCAP functioned, the original host of the OCAP server accepted map requests, and he would then generate map tile data for use when rendering the mission data. We intend to provide an example of this type, as well as the method to generate map data yourself, so that anyone who wants to host or use OCAP can play on any maps they want to. In addition, after the mission is complete, the a3 server mod outputs a series of JSON files that need to be placed in the OCAP server for it to parse as an operation. We could make a file selector to allow the a3 server host to select their a3 directory, so that OCAP can just automatically suck in the mission files after they are generated. 

Also, with the current self-contained webapp, it is possible to rip everything out of the www directory, and host it from the same PC as the a3 server with your own web server apps. For prototyping, we just had local Apache servers set up. I'm hesitant to package it that way by default, as I wouldn't want to provide an implementation that might not be secured properly, but we could provide users with some basic guidance to set that up for their groups. That way, the default is a standalone for anyone who wants it to be simple and safe for personal or small group use, but for those who want to host it for large groups, or want a single URL, they can pull the files into their own web server and connect it up to their a3 server to host to a group. Once hosted, all users should be able to view any operations that the host has pulled into OCAP. Getting users to view a synchronized broadcast is a far more difficult hurdle, as the application at its base isn't designed to support it. We would need to implement that from scratch, and that would be a long term effort, but if the interest is there, I'll consider it.

TL:DR the current implementation can be ripped as-is to your own web server to be hosted from the same PC as your a3 server, and we can add the ability for the user to pick their a3 server directory so that the server can find the mission files. We could potentially automate the system for map generation to make it easier, and provide some guidance for setting up a locally hosted web server, to use at your own risk. We could also add as a long-term goal the mechanisms to allow for synchronized broadcasts to users, but that would involve significant redesign.

Would that be what you are looking for? Those are very intelligent questions, don't sell yourself short! 😁

**Comment by schbaeggle** [comment]
> to be hosted from the same PC as your a3 server

Which excludes everyone using a non-root-server :confused:

So if I understand correctly, the dataflow would be either manual or via local files. How can a client download the data? Wouldn't it be cool to have your app at least quickly connect via ftp and fetch that json (or whatever format) data by itself. That way, one can simply setup a cronjob to automate it or have a neat little button to manually suck up data and render them.
(brainstorming: what about feeding the data into a mysql database (via extDB), localhost or any other host instead of dealing with files? Thereafter your app can pull them and render them again in whatever form..)

As most groups already got some kind of web-based system running, a simple "upload-install-run" (like you do on almost every web-app you'd like to install) solution fits quite good, this is why I did not see the actual use of a packed standalone in the first place ;) Based on what webserver you own, you can dig into automating it yourself.

Btw, using OCAP for a synced playback or even whiteboard was just a visionary dream after all ;) I know it's not impossible, I think there is a system for synchronized youtube videos (https://www.watch2gether.com), I guess websocket is the way to go there. This is, where my experience ends. But as you sad, this basically means a complete rework of that app. Lots and lots of work :grin:

So if I can just extract the webapp to a dedicated webserver and have it pull the json files from my a3 server, I'd be more than happy. I really can't wait 😁

**Comment by TelluriumCrystal** [comment]
First of all lets make sure we all clearly understand how OCAP Revived will work:

1. User adds OCAP recorder module to a mission and specifies export directory (directory is set either in the module settings or in the CBA settings)
2. User runs mission and OCAP addon records data
3. User triggers OCAP export through one of several methods
4. OCAP exporter builds mission track JSON from temporary file(s) and saves it to the directory specified in step 1
5. User opens file in OCAP viewer

My understanding of your situation is that you have an Arma 3 server running on one machine, and a web server running on another. You want to be able to have users access the OCAP viewer on your web server and view missions there as opposed to having users get the JSON themselves and view it with a local copy of the OCAP viewer. This should be possible with almost no modifications to the system.

You should be able to install the OCAP viewer onto your web server without much difficulty (I will have better instructions for this in the wiki that I will eventually get set up). The viewer itself is written in Javascript and can simply be copied to the document tree of your server. On the Arma 3 server you can specify a filepath you want to save all your mission tracks to. Then all you need to do is set up a means of automatically sending files from that folder to the ocap/data folder on your web server. There are all kinds of ways to do this and I will leave that up to you to set up.

I am open to adding FTP export capability to OCAP but that won't get implemented until after 0.7.0.

Am I correctly understanding your situation and what are your thoughts on this solution?

**Comment by schbaeggle** [comment]
You nailed your description of my situation. I assume that this is the case for a wide range of players/groups below a certain amount of players.
As long as it's as easy to set up onto a dedicated webserver as you described (or will be describing in the future), everyone will be happy. Downloading that recorded files via FTP should not be a big deal for everyone familiar with linux, cronjobs etc. However, it'll be a neat feature for your OCAP, dedicated-server-version as well as local version in the future for sure.

I'm looking forward for your first release! 😁
